### PR TITLE
Add support for caveats in LookupSubjects API

### LINKED ISSUE #931: Add support for caveats in LookupSubjects API

### PR DESCRIPTION
**Pull request by josephschorr**
Fixes #931

### CODE REVIEW

File: internal/util/subjectsetbytype_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+func RR(namespaceName string, relationName string) *core.RelationReference {
+	return &core.RelationReference{
+		Namespace: namespaceName,
+		Relation:  relationName,
+	}
+}
```
Comment (vroldanbet) [inline internal/util/subjectsetbytype_test.go:18]: this method is duplicated quite a lot in the codebase, especially in tests. I understand the need for making the tests readable with a short method name, but I think we should at least implement it in terms of `pkg/namespace/builder.go:RelationReference`

File: internal/util/subjectsetbytype.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// Len returns the number of keys in the set.
+func (s *SubjectByTypeSet) Len() int {
+	return len(s.byType)
+}
+
+// SubjectSetForType returns the subject set associated with the given subject type, if any.
+func (s SubjectByTypeSet) SubjectSetForType(rr *core.RelationReference) (SubjectSet, bool) {
+	typeKey := fmt.Sprintf("%s#%s", rr.Namespace, rr.Relation)
+	found, ok := s.byType[typeKey]
+	return found, ok
```
Comment (vroldanbet) [inline internal/util/subjectsetbytype.go:97]: also this method is untested by the unit test

File: internal/util/subjectsetbytype.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+func wrapCaveat(caveat *core.ContextualizedCaveat) *v1.CaveatExpression {
+	if caveat == nil {
+		return nil
+	}
+
+	return &v1.CaveatExpression{
+		OperationOrCaveat: &v1.CaveatExpression_Caveat{
+			Caveat: caveat,
+		},
+	}
```
Comment (vroldanbet) [inline internal/util/subjectsetbytype.go:109]: untested by the unit test -> I think we could modify the unit test to have a caveated tuple in place, just to make these bits are correctly handled

File: internal/util/subjectsetbytype_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+func TestSubjectByTypeSet(t *testing.T) {
+	assertHasObjectIds := func(s *SubjectByTypeSet, rr *core.RelationReference, expected []string) {
+		wasFound := false
+		s.ForEachType(func(foundRR *core.RelationReference, subjects SubjectSet) {
+			objectIds := make([]string, 0, len(subjects.AsSlice()))
+			for _, subject := range subjects.AsSlice() {
+				require.Nil(t, subject.GetCaveatExpression())
+				require.Empty(t, subject.GetExcludedSubjects())
```
Comment (vroldanbet) [inline internal/util/subjectsetbytype_test.go:27]: let's add caveats to the subjects so we test SubjectByTypeSet handles caveated subjects

File: internal/caveats/run.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+		if childResult.IsPartial() {
+			return childResult, nil
+		}
+
+		switch cop.Op {
+		case v1.CaveatOperation_AND:
+			boolResult = boolResult && childResult.Value()
+
+			if debugOption == RunCaveatExpressionWithDebugInformation {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/caveats/run.go:163]: The tests are in my other branch

File: internal/util/subjectsetbytype.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// AddConcreteSubject adds a non-caveated subject to the set.
+func (s *SubjectByTypeSet) AddConcreteSubject(subject *core.ObjectAndRelation) {
+	s.AddSubject(subject, nil)
+}
+
+// AddSubject adds the specified subject to the set.
+func (s *SubjectByTypeSet) AddSubject(subject *core.ObjectAndRelation, caveat *core.ContextualizedCaveat) {
+	typeKey := fmt.Sprintf("%s#%s", subject.Namespace, subject.Relation)
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/util/subjectsetbytype.go:38]: There is a TODO for that

File: internal/util/subjectsetbytype_test.go (hunk 1 above)
Comment (josephschorr) [reply internal/util/subjectsetbytype_test.go:18]: its a tradeoff; some tests we do `var RR = tuple.RR` or somesuch, but either way, it should not be changed in this PR

File: internal/caveats/run.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+func combineMaps(first map[string]any, second map[string]any) map[string]any {
+	if first == nil {
+		first = make(map[string]any, len(second))
+	}
+
+	cloned := maps.Clone(first)
+	maps.Copy(cloned, second)
+	return cloned
+}
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/caveats/run.go:223]: This only happens when debug is enabled, so I really don't think it matters

File: internal/util/subjectsetbyresourceid.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	ssr.subjectSetByResourceID[resourceID].Add(subject)
+}
+
+// AddFromRelationship adds the subject found in the given relationship to this map, indexed at
+// the resource ID specified in the relationship.
+func (ssr SubjectSetByResourceID) AddFromRelationship(relationship *core.RelationTuple) {
+	ssr.add(relationship.ResourceAndRelation.ObjectId, &v1.FoundSubject{
+		SubjectId:        relationship.Subject.ObjectId,
+		CaveatExpression: wrapCaveat(relationship.Caveat),
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/util/subjectsetbyresourceid.go:41]: Okay

File: internal/util/subjectsetbytype.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	return len(s.byType) == 0
+}
+
+// Len returns the number of keys in the set.
+func (s *SubjectByTypeSet) Len() int {
+	return len(s.byType)
+}
+
+// SubjectSetForType returns the subject set associated with the given subject type, if any.
+func (s SubjectByTypeSet) SubjectSetForType(rr *core.RelationReference) (SubjectSet, bool) {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/util/subjectsetbytype.go:94]: Whoops

**Comment by josephschorr** [comment]
Updated

File: internal/caveats/run.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/authzed/spicedb/pkg/caveats"
+	"github.com/authzed/spicedb/pkg/datastore"
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+)
+
+type RunCaveatExpressionDebugOption int
+
+const (
+	RunCaveatExpressionNoDebugging          RunCaveatExpressionDebugOption = 0
+	RunCaveatExpressionWithDebugInformation RunCaveatExpressionDebugOption = 1
```
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/caveats/run.go:24]: I'm not sure I follow. If this flag is not being used, how is that internal library triggering the codepath?

File: internal/caveats/run.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/caveats/run.go:228]: that is true. But it's a free optimization, right? :D

File: internal/caveats/run.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/caveats/run.go:168]: if the branch is pushed, do you mind linking it here?

File: internal/graph/lookupsubjects.go [hunk 11]
```diff
@@ -317,18 +330,71 @@ func (cl *ConcurrentLookupSubjects) dispatchTo(
...
-				Metadata:      addCallToResponseMetadata(result.Metadata),
-			}, true, nil
-		},
-	}
+	toDispatchByType.ForEachType(func(resourceType *core.RelationReference, foundSubjects util.SubjectSet) {
+		slice := foundSubjects.AsSlice()
+		resourceIds := make([]string, 0, len(slice))
+		for _, foundSubject := range slice {
+			resourceIds = append(resourceIds, foundSubject.SubjectId)
+		}
```
Comment (vroldanbet) [inline internal/graph/lookupsubjects.go:338]: seems like something that could be added as a method in `SubjectSet`?

**Review by vroldanbet — APPROVED** [review]

File: internal/services/v1/permissions.go [hunk 12]
```diff
@@ -427,16 +430,47 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
...
+		foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+		if !ok {
+			return fmt.Errorf("missing resource ID in returned LS")
+		}
+
+		for _, foundSubject := range foundSubjects.FoundSubjects {
 			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
 			for _, excludedSubject := range foundSubject.ExcludedSubjects {
 				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
 			}
```
Comment (vroldanbet) [inline internal/services/v1/permissions.go:442]: We err on the side of safety by keeping the deprecated excluded subject IDS with a conservative value that considers subjects for which caveat evaluation missed context as excluded.

**Review by vroldanbet — APPROVED** [review]

**Review by vroldanbet — APPROVED** [review]

**Pull request merged**